🤡c🤡f🤡-🤡1🤡1🤡8🤡1🤡D 🤡 (#3909)
cpinitiative/usaco-guide: solutions/gold/cf-1181D.mdx

---
id: cf-1181D
source: CF
title: Irrigation
author: Chongtian Ma
---

[Official Analysis](https://codeforces.com/blog/entry/67727)

## Explanation

> We can subtract $n$ from each of the queries, and answer them offline. Then, sort the cities by occurrences and queries by year.

Reviewer: ```suggestion
We can subtract $n$ from each query, answer them offline, and then sort the cities by occurrences and the queries by year.
```

Author: thats not the right order bro
you should sort before u answer queries offline

Reviewer: in ur initial sentences it isn't clear, so I suggest you fix it

> Let city $i$ denote the $i$'th smallest city sorted by occurrences, and $occ_i$ denote the number of occurrences. We can loop  through each $i$ and track the number of years has gone by. Assume cities $1, 2, ...i$ already have $occ_i$ occurrences. For them to all end up with $occ_{i+1}$ occurrences, we need $i \cdot (occ_{i+1} - occ_i)$ more years. This means that for the next $i \cdot (occ_{i+1} - occ_i)$ years, cities $1, 2, ... i$ will be selected one after another.

Reviewer: ```suggestion
Let city $i$ denote the $i$'th smallest city sorted by occurrences, and $occ_i$ represent the number of occurrences. We can loop through each $i$ and track the years that have passed. 

Assume cities $1, 2, ...i$ already have $occ_i$ occurrences. For them to all end up with $occ_{i+1}$ occurrences, we need $i \cdot (occ_{i+1} - occ_i)$ more years, meaning that for the next $i \cdot (occ_{i+1} - occ_i)$ years, cities $1, 2, ... i$ will be selected one after another.
```

> How do we answer individual years? We can use a BST to track the initial indicies of each city $1, 2, 3... i$. Note that tie for occurrences is broken by initial indices, not the sorted.

Reviewer: ```suggestion
To answer individual years, we can use a BST to track the initial indices of each city $1, 2, 3... i$. Note that initial indices, not the sorted ones, break the tie for occurrences.
```

Reviewer 2: use \dots

> Let the current year be $y$, then at each iteration we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$. We can use binary search to obtain these queries. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indicies. Since we have been tracking initial indicies with our BST, we can simply perform an index query for this information.

Reviewer: ```suggestion
Let the current year be $y$. At each iteration, we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$. We can use binary search to obtain these queries. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indices. 
```

> There might be some leftover queries that ask for years after the final $y$. At this point, we know all cities have the max occurrence, so the answer to them is just $(q - y) \% m$.

Reviewer: ```suggestion
At this point, we know all cities have the max occurrence, so the answer to remaining queries that ask for years after the final $y$ is just $(q - y) \% m$.
```

## Implementation

**Time Complexity:** $\mathcal{O}(m \log q \log m)$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/detail/standard_policies.hpp>
#include <ext/pb_ds/tree_policy.hpp>
using namespace std;
using namespace __gnu_pbds;

using ll = long long;
#define all(x) x.begin(), x.end()

template <typename T>
using ordered_set =
    tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;

int main() {
	int n, m, q;
	scanf("%d%d%d", &n, &m, &q);
	vector<int> counter(m);
	for (int i = 0; i < n; i++) {
		int x;

> 		scanf("%d", &x);

Reviewer 2: me when no cin

Author: ryan when no cin?

---
id: cf-1181D
source: CF
title: Irrigation
author: Chongtian Ma
---

[Official Analysis](https://codeforces.com/blog/entry/67727)

## Explanation

> We can subtract $n$ from each of the queries, sort the cities by occurrences and queries by year, then answer the queries offline.

Reviewer: ```suggestion
We can subtract $n$ from each query, sort the cities by occurrences and the queries by year, and then answer the queries offline.
```

Let city $i$ denote the $i$'th smallest city sorted by occurrences, and $occ_i$ represent the number of occurrences. We can loop through each $i$ and track the years that have passed.

Assume cities $1, 2, \dots i$ already have $occ_i$ occurrences. For them to all end up with $occ_{i+1}$ occurrences, we need $i \cdot (occ_{i+1} - occ_i)$ more years, meaning that for the next $i \cdot (occ_{i+1} - occ_i)$ years, cities $1, 2, \dots i$ will be selected one after another.

> To answer individual years, we can use a BST to track the initial indices of each city $1, 2, 3 \dots i$. Note that initial indices, not the sorted ones, break the tie for occurrences.

Reviewer: For this part "we can use a BST to track the initial indices", can you link the module for BST

Author: ???? this is in the BST module lol???

Reviewer: oh shoot I just looked over the writing but ok disregard my comment

> Let the current year be $y$. At each iteration, we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$. We can use binary search to obtain these queries. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indices.

Reviewer: "We can use binary search to obtain these queries." doesn't really make sense

> Let the current year be $y$. At each iteration, we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$. We can use binary search to obtain the indices of these queries. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indices.

Reviewer: ```suggestion
Let the current year be $y$. At each iteration, we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$ using binary search. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indices.
```

Reviewer: does this work?

Author: No, you are using binary search to OBTAIN the queries, not answer them

> At this point, we know all cities have the max occurrence, so the answer to remaining queries that ask for years after the final $y$ is just $(q - y) \% m$.

Reviewer: ```suggestion
At this point, we know every city has the max occurrence, so the answer to remaining queries that ask for years after the final $y$ is just $(q - y) \% m$.
```

How do we answer individual years? We can use a BST to track the initial indicies of each city $1, 2, 3... i$. Note that tie for occurrences is broken by initial indices, not the sorted.

Let the current year be $y$, then at each iteration we can answer every query $q$ in the range of $y \leq q < y + i \cdot (occ_{i+1} - occ_i)$. We can use binary search to obtain these queries. Since these cities are selected one after another, the answer to query $q$ is just the $(q - y) \% i$'th city sorted by initial indicies. Since we have been tracking initial indicies with our BST, we can simply perform an index query for this information.

There might be some leftover queries that ask for years after the final $y$. At this point, we know all cities have the max occurrence, so the answer to them is just $(q - y) \% m$.

## Implementation

**Time Complexity:** $\mathcal{O}(m \log q \log m)$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
#include <ext/pb_ds/tree_policy.hpp>
#include <ext/pb_ds/detail/standard_policies.hpp>
using namespace std;
using namespace __gnu_pbds;

using ll = long long;

> #define all(x) x.begin(), x.end()

Author: literlaly everyone uses all(x)

template<typename T>
using ordered_set = tree<T, null_type, less<T>, rb_tree_tag, tree_order_statistics_node_update>;

int main(){
	int n, m, q;
	scanf("%d%d%d", &n, &m, &q);
	vector<int> counter(m);
	for(int i = 0; i < n; i++){
		int x;
		scanf("%d", &x);
		counter[--x]++;
	}
	vector<pair<ll, int>> query(q); // {year, index}
	for(int i = 0; i < q; i++){
		scanf("%lld", &query[i].first);
		query[i].first -= n;
		query[i].second = i;
	}
	sort(all(query));
	vector<pair<ll, int>> occ(m); // {occurrence, city}
	for(int i = 0; i < m; i++){
		occ[i] = {counter[i], i};
	}
	sort(all(occ));
	vector<int> qans(q, -1);
	// tracks the initial indicies of the cities so far in sorted order
	ordered_set<int> track_city;
	ll years = 1;

> 	for(int i = 0; i < m - 1; i++){

Author: why isnt the bot autoformatting???????

		ll years_upper = years + (occ[i+1].first - occ[i].first) * (i + 1);
		track_city.insert(occ[i].second);
		int query_low_index = lower_bound(all(query), make_pair(years, 0)) - query.begin();
		for(int j = query_low_index; j < q && query[j].first < years_upper; j++){
			ll qyears = query[j].first;
			int qindex = query[j].second;
			qans[qindex] = *track_city.find_by_order((qyears - years) % (i + 1));
		}

> 		years = years_upper;

Author: i explained this exact thing in explanation, idk what to add for comments here

> 		years = years_upper;

Reviewer 3: reiterate it
if you need some help look at daniel's comments
